### ISSUE TITLE
Reproduce Results on CUB dataset

### ISSUE THREAD

**Issue by yangyu12** [anvoynov/BigGANsAreWatching#2]
Hi, remarkable work! Congratulations! I am trying to reproduce the segmentation results on CUB dataset. However, I got performance significantly different what is claimed in Table 1. Could you please help me to see if I did something wrong? Or if this result is within the normal range of variation? Many Thanks!

### Reproduce results
![image](https://user-images.githubusercontent.com/17905452/90380306-d05f7880-e073-11ea-8060-8f6226a864c6.png)

### Commands
BigBiGAN
`python train_segmentation.py --out=results/CUB_BigBiGAN --gen_devices=1 --gan_weights=BigGAN/weights/BigBiGAN_x1.pth --z=embeddings/BigBiGAN_ImageNet_z.npy --bg_direction=BigGAN/weights/bg_direction.pth --val_images_dir=data/CUB/test_images --val_masks_dir=data/CUB/test_segmentations`

E-BigBiGAN (w/o z-noising)
`python train_segmentation.py --out=results/CUB_E_BigBiGAN --gan_weights=BigGAN/weights/BigBiGAN_x1.pth --z=embeddings/BigBiGAN_CUB_train_z.npy --bg_direction=BigGAN/weights/bg_direction.pth --val_images_dir=data/CUB/test_images --val_masks_dir=data/CUB/test_segmentations`

E-BigBiGAN (w/ z-noising)
`python train_segmentation.py --out=results/CUB_E_BigBiGAN_noise --gen_devices=1 --gan_weights=BigGAN/weights/BigBiGAN_x1.pth --z=embeddings/BigBiGAN_CUB_train_z.npy --z_noise=0.2 --bg_direction=BigGAN/weights/bg_direction.pth --val_images_dir=data/CUB/test_images --val_masks_dir=data/CUB/test_segmentations`

For the `val_images_dir` and `val_masks_dir`, I just copied all the test images and segmentations to new folders, where test split follows ReDO (Chen et al., 2019) setting.

### Environment
Ubuntu 18.04.4 LTS
GeForce GTX TITAN X
torch==1.3.0
torchvision==0.4.1a0+d94043a
scipy==1.3.1
numpy==1.17.3
Pillow==6.2.1
h5py==2.9.0
scikit_image==0.15.0

**Comment by anvoynov** [comment]
Hi @yangyu12,
though we report the avg metrics, all of them have low variance with respect to seed so this is strange.
Could you please attach the args.json file generated in the output directory? Is that the numbers from the score.json?
It would be super-useful if you also could share the model weights (for E-BigBiGAN (w/ z-noising)).

**Comment by yangyu12** [comment]
By the way, I have some confusion about the evaluation code. During training, the model output 2-channel masks and cross entropy loss is computed. However, in inference mode, only the second output channel is taken, and a sigmoid function is used to turn the logits into probabilities. Is it more proper to apply softmax over channels instead of the sigmoid function?  
(c.f. https://github.com/anvoynov/BigGANsAreWatching/blob/master/postprocessing.py#L46)

Another issue is that at https://github.com/anvoynov/BigGANsAreWatching/blob/master/metrics.py#L26 and https://github.com/anvoynov/BigGANsAreWatching/blob/master/metrics.py#L29 when mask_area is 0.0, shouldn't the precision or recall be 0.0? I know this rarely happens but this code seems problematic.

**Comment by yangyu12** [comment]
Hi @anvoynov, sure, the content of args.json file for E-BigBiGAN (w/ z-noising) is as following:

{"args": null, "out": "results/CUB_E_BigBiGAN_noise", "gan_weights": "BigGAN/weights/BigBiGAN_x1.pth", "bg_direction": "BigGAN/weights/bg_direction.pth", "device": 0, "gen_devices": [1], "seed": 2, "z": "embeddings/BigBiGAN_CUB_train_z.npy", "z_noise": 0.2, "val_images_dirs": ["data/CUB/test_images"], "val_masks_dirs": ["data/CUB/test_segmentations"], "rate": null, "steps_per_rate_decay": null, "rate_decay": null, "n_steps": null, "latent_shift_r": null, "batch_size": null, "steps_per_log": null, "steps_per_checkpoint_save": null, "steps_per_validation": null, "test_samples_count": null, "synthezing": null, "mask_size_up": null, "connected_components": null, "maxes_filter": null}

I just checked the score.json, the number is the same.

I have packed the output folder of E-BigBiGAN (w/ z-noising) and dropped it here: 
https://drive.google.com/file/d/1R1jdy1URxjIEnQJYwo71gtlZYXapNyfy/view?usp=sharing

Thanks for help!

**Comment by anvoynov** [comment]
@yangyu12 thank you so much for pointing on it! The softmax is indeed is the correct option there. This has happened once I have transferred and refactored my local code for git, will fix it soon (the reported numbers are correct). Though this seems not to be the reason of the mismatch as it practically does not affect the metrics. Is there a chance there is no file embeddings/BigBiGAN_CUB_train_z.npy?

**Comment by yangyu12** [comment]
@anvoynov Yes, you are right. It doesn't affect the metrics too much. Metrics even decrease a very little bit when I use softmax.  

I have checked my folder. The embedding files are indeed there. It is downloaded from the URL you provided (I just run the download script). I wonder if it is possible that you offer the wrong embedding. Notice that in my reproducing results training with ImageNet embedding (BigBiGAN) is even better than CUB_train embedding (E-BigBiGAN with z-noising), it is strange.

**Comment by yangyu12** [comment]
Oh, sorry. I seem to make a mistake. In your Table 1, BigBiGAN means training on synthetic data generated from noise instead of ImageNet embedding?
However, the results are still strange. I still haven't figured out where is wrong.

**Comment by anvoynov** [comment]
@yangyu12 Oh, finally I discovered the problem. Once I have refactored the code, I have missed the ```latent_shift_r``` multiplier:
https://github.com/anvoynov/BigGANsAreWatching/blob/master/gan_mask_gen.py#L82
fixed in this commit:
https://github.com/anvoynov/BigGANsAreWatching/commit/d695f08fbf6e2918956722c4e513e703907463b6

right now I have launched the training with exactly this code, hopefully, all would be alright and I will share logs and the model here. Once again, thank you very much for pointing on it!

**Comment by anvoynov** [comment]
@yangyu12 now everything should work fine.
Please check the model and logs produced by the training with your command:
https://www.dropbox.com/s/lcex44joxz6uo3o/CUB_E_BigBiGAN_noise.tar?dl=0

On CUB-test it has:
'IoU': 0.716
'accuracy': 0.940
'F_max': 0.836

**Comment by yangyu12** [comment]
Hi, @anvoynov Thanks very much for the help. I have trained the model with your updated code got a result:

F_max: 0.810, IoU: 0.684, accuracy: 0.944

I think this is close and the gap may be due to different environment. Thanks, after all.

**Comment by anvoynov** [comment]
@yangyu12 thank you for sharing! Could you please upload the command.sh, args.json, and the final model?
The gap steel looks a bit strange though it could be some env issues indeed.

**Comment by yangyu12** [comment]
@anvoynov Just wait for me to switch to pytorch 1.4.0 & torchvision 0.5.0 and rerun the experiment. I will update the results and share the full folder here once I finish.

**Comment by yangyu12** [comment]
Hi @anvoynov, I have checked the synthetic data, but some of them don't look like birds at all. See the following examples. I don't know if it is normal.
I am generating the embedding for the training set myself and seeing if it fixes the problem. 
![image](https://user-images.githubusercontent.com/17905452/90800327-6b727f80-e30c-11ea-8986-25681d65985e.png)
(1st row: original image.  2nd row: decoded image from embedding.  3rd row: moved along bg direction.  4th row: the masks.)

**Comment by yangyu12** [comment]
This is the results from my generated embeddings. I can't say they are perfect but seems more reasonable. I will see if it fixes the problem by training on my generated embeddings.

![image](https://user-images.githubusercontent.com/17905452/90803522-c7d79e00-e310-11ea-905e-e9637ac237e9.png)

**Comment by anvoynov** [comment]
> Note:
> This synthetic data is generated by decoding the provided embedding `BigBiGAN_CUB_train_z.npy` with the BigBiGAN generator. No noise is introduced so the results should be deterministic. I assume that the embedding observes the order that the training images observe, thus the visualization is realised.

I'm not quite sure that this is true as the `UnannotatedDataset` used in the embeddings computation may sort the files in a different way. Though the `BigBiGAN_CUB_train_z.npy` contains all the embeddings, the inner permutation may vary. I just checked the embeddings reconstructions, it looks fine actually:
![image](https://user-images.githubusercontent.com/34134350/90989076-77766f80-e5a0-11ea-8db0-d3adc213ef80.png)
that's why I don't think this is the issue of the results mismatch in your run. Could you please share the outputs of your run?

**Comment by yangyu12** [comment]
You are right. The gap has nothing to do with the embedding. I have tried following things, all of them give me very close results to
{"F_max": 0.809025228023529}, "thr": {"IoU": 0.684075553198345, "accuracy": 0.9438497857451439}}

-  Switch to a new environment and meet `requirements.txt`. Results are available: https://drive.google.com/file/d/1gnncucb5nfmWkjLtJFHJpiTaos6BzicY/view?usp=sharing

- Use my generated embedding. Results are very close, omitted.

And I have evaluated the following model in my environment, 

> @yangyu12 now everything should work fine.
> Please check the model and logs produced by the training with your command:
> https://www.dropbox.com/s/lcex44joxz6uo3o/CUB_E_BigBiGAN_noise.tar?dl=0

it gives me 
```
Segmenation model {'F_max': 0.8077686429023743}
Segmenation model {'IoU': 0.6836832139976322, 'accuracy': 0.9437406770288944}
```

It's weird. Could it be possible that there is some error in the evaluation code? Can you provide your `train_val_test_split.txt` as well? Just make sure that we are testing on the same test set.

B.T.W. the exact code I use is https://github.com/yangyu12/BigGANsAreWatching

**Comment by anvoynov** [comment]
I'm out of my code until Monday, sorry. Just in case: do you apply square crop + resize to 128, like in the ReDo paper?

**Comment by yangyu12** [comment]
I didn't do square crop. I only did resize. See https://github.com/yangyu12/BigGANsAreWatching/blob/master/eval_segmentation.py#L50, where `crop=False`.

**Comment by anvoynov** [comment]
Ah, sorry for the confusing flag here. This code is actually fit for ECSSD and others SOD datasets so the birds should be evaluated with `crop=True` for fare compare with the ReDo (or the images files should be preprocessed with the crop). I should certainly add this flag to the args.

**Comment by yangyu12** [comment]
Thanks! 
After setting `crop=True`, evaluating your shared model gives me
```
Segmenation model {'F_max': 0.8364795446395874}
Segmenation model {'IoU': 0.7165820255889558, 'accuracy': 0.9403773924410344}
```

Evaluating my trained model `CUB_E_BigBiGAN_noise` gives me
```
Segmenation model {'F_max': 0.8388442397117615}
Segmenation model {'IoU': 0.7189271003338509, 'accuracy': 0.9408648254275322}
``` 

Evaluating my model trained on my generated embeddings gives me
```
Segmenation model {'F_max': 0.8429229259490967}
Segmenation model {'IoU': 0.7220172818554565, 'accuracy': 0.9404425908029079}
```

Evaluating my model trained with pytorch==1.4.0 gives me
```
Segmenation model {'F_max': 0.8401502966880798}
Segmenation model {'IoU': 0.7195895961956121, 'accuracy': 0.9409913717508316}
```

Therefore, I think this issue is solved. The performance are quite stable. And I think there is actually no pytorch version issue. This code works quite well with pytorch==1.3.0.
